### ISSUE TITLE
Errors in subsequent requests caused by connection reuse, 204 responses combined with chunked Transfer Encoding

### ISSUE THREAD

**Issue by bauglir** [JuliaWeb/HTTP.jl#261]
I initially encountered this issue in my testing framework based on HTTP.jl when testing `DELETE` requests where the server returns a `204` response. These specific test cases would pass properly, but the subsequent test would fail.

From further investigation I believe I have narrowed it down to being unrelated to `DELETE` requests and the issue having to do with the way `204` responses are send over the pipe when using chunked transfer encoding. It could be this problem would arise with any response without a body, I haven't been able to verify that at this point in time.

When a request receives a `204` response it appears to get handled properly. However, an `excess` is left on the connection which is parsed and and causes a crash on the subsequent request on the same connection. This is reported as being due to an invalid status line being encountered. Instead the sequence `0\r\n\r\n` is parsed from the excess, which I believe indicates this is related to chunked transfer encoding.

The issue is easily reproduced by starting a server similar to the standard `serve` implementation including `handle_request`. The following code reproduces the issue in its most dressed down form I've been able to get it
```julia
using HTTP

@async HTTP.listen() do http::HTTP.Stream
  request::HTTP.Messages.Request = http.message
  request.response = HTTP.Messages.Response(204)

  HTTP.startwrite(http)
  write(http, request.response.body)
end

sleep(1)

response = HTTP.request("GET", "http://localhost:8081")
println(response)
response = HTTP.request("GET", "http://localhost:8081")
println(response)
```

This results in
```bash
$ julia --color=yes ./204_errors.jl                                                                                                                                                                                                  130 ↵
I- Listening on: Base.InetAddr{IPv4}(ip"127.0.0.1", 0x1f91)
I- Accept:  🔗    0↑     0↓    0s 127.0.0.1:8081:8081 ≣16
HTTP.Messages.Response:
"""
HTTP/1.1 204 No Content
Transfer-Encoding: chunked

"""
ERROR: LoadError: W- HTTP.Parsers.ParseError(:INVALID_STATUS_LINE, "0\r")write: broken pipe (EPIPE)
while loading /.../204_errors.jl, in expression starting on line 15
```

Forcing the requests to not reuse existing connections works fine, as expected.

I'm afraid this is about as far as I can parse the issue and I am not entirely sure whether this is a 'server-side' or 'client-side' issue in HTTP.jl. As clients such as Postman and cURL (see the included log) do not seem to be affected when talking to the server implementation mentioned earlier, I expect this to be on the 'client-side'. But it could also be that the server simply shouldn't be sending the chunked transfer encoding information in case there is no body in a response.

## cURL Log
```bash
$ curl -o curl-log -v localhost:8081 localhost:8081
* Rebuilt URL to: localhost:8081/
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
Dload  Upload   Total   Spent    Left  Speed
0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8081 (#0)
> GET / HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.58.0
> Accept: */*
>
< HTTP/1.1 204 No Content
< Transfer-Encoding: chunked
<
{ [5 bytes data]
0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Connection #0 to host localhost left intact
* Rebuilt URL to: localhost:8081/
* Found bundle for host localhost: 0x559e857485b0 [can pipeline]
* Re-using existing connection! (#0) with host localhost
* Connected to localhost (127.0.0.1) port 8081 (#0)
> GET / HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.58.0
> Accept: */*
>
< HTTP/1.1 204 No Content
< Transfer-Encoding: chunked
<
* Connection #0 to host localhost left intact
```
I do find it interesting how only the first request in this case appears to include the additional chunked transfer encoding bytes. I have tried with even more requests and in those cases it's still only the first response containing the additional data.

### Versions used
Julia 0.6.3
HTTP.jl 0.6.12
MbedTLS.jl 0.5.11

**Comment by quinnj** [comment]
Thanks for the detailed report; your guess sounds likely; that we're not fully finishing parsing those 204 responses. @samoconnor and I should be able to figure out what's going on here.

**Comment by samoconnor** [comment]
Regarding the `listen` example...
```julia
@async HTTP.listen() do http::HTTP.Stream
  request::HTTP.Messages.Request = http.message
  request.response = HTTP.Messages.Response(204)

  HTTP.startwrite(http)
  write(http, request.response.body)
end
```

It seems not quite right to overwrite the `request.response` with a new Response object, it will probably work ok, but the back-reference from the Response to the Request won't be set properly.

One of the following might be better:
```julia
@async HTTP.listen() do r::HTTP.Request
    return HTTP.Response(204)
end

@async HTTP.listen() do http::HTTP.Stream
    HTTP.setstatus(http, 204)
    HTTP.startwrite(http)
end
```

**Comment by samoconnor** [comment]
> I do find it interesting how only the first request in this case appears to include the additional chunked transfer encoding bytes. I have tried with even more requests and in those cases it's still only the first response containing the additional data.

I assume you're referring to the fact that the curl log only says `[5 bytes data]` the first time. I think this is just a curl thing. When I specify a `-o` for both requests the output looks like this:

```bash
$ curl -o curl-log -v localhost:8081 -o curl-log2 -v localhost:8081
* Rebuilt URL to: localhost:8081/
-- snip --
< Transfer-Encoding: chunked
<
{ [5 bytes data]
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
* Connection #0 to host localhost left intact
-- snip --
< Transfer-Encoding: chunked
<
{ [5 bytes data]
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Connection #0 to host localhost left intact
```

... and in wireshark I see this:

```bash
curl -o curl-log -v localhost:8081  localhost:8081 localhost:8081
```
```http
GET / HTTP/1.1
Host: localhost:8081
User-Agent: curl/7.54.0
Accept: */*

HTTP/1.1 204 No Content
Transfer-Encoding: chunked

0

GET / HTTP/1.1
Host: localhost:8081
User-Agent: curl/7.54.0
Accept: */*

HTTP/1.1 204 No Content
Transfer-Encoding: chunked

0

GET / HTTP/1.1
Host: localhost:8081
User-Agent: curl/7.54.0
Accept: */*

HTTP/1.1 204 No Content
Transfer-Encoding: chunked

0
```

So, it looks like the server side is consistently sending out `0\r\n\r\n` to indicate end of chunked body.

**Comment by samoconnor** [comment]
It looks like we're not expecting any bytes after the header for a 204 response because the RFC says "All 1xx (Informational), 204 (No Content), and 304 (Not Modified)
 responses do not include a message body." https://tools.ietf.org/html/rfc7230#section-3.3 

The function [`Messages.bodylength`](https://github.com/JuliaWeb/HTTP.jl/blob/master/src/Messages.jl#L280) returns `0` for 204 Responses.

The patch below changes `bodylength` to return `unknown_length` for 204 messages with `Transfer-Encoding: chunked`. This should cause the HTTP.jl client to consume the `0\r\n\r\n` at the end of the zero length chunked body.

```patch
--- a/src/Messages.jl
+++ b/src/Messages.jl
@@ -279,6 +279,7 @@ headerscomplete(r::Request) = r.method != ""
 """
 bodylength(r::Response)::Int =
                  r.request.method == "HEAD" ? 0 :
+                               ischunked(r) ? unknown_length :
                      r.status in [204, 304] ? 0 :
     (l = header(r, "Content-Length")) != "" ? parse(Int, l) :
                                               unknown_length
```

Please try this out and let us know if it works for you.

**Comment by samoconnor** [comment]
The following patch ensures that the server does not return a `Transfer-Encoding: chunked` header for messages that have no body (i.e. HEAD responses, 204 and 304).

```patch
--- a/src/Streams.jl
+++ b/src/Streams.jl
@@ -72,7 +72,8 @@ function IOExtras.startwrite(http::Stream)
     m = messagetowrite(http)
     if !hasheader(m, "Content-Length") &&
        !hasheader(m, "Transfer-Encoding") &&
-       !hasheader(m, "Upgrade")
+       !hasheader(m, "Upgrade") &&
+        bodylength(m) > 0
         http.writechunked = true
         setheader(m, "Transfer-Encoding" => "chunked")
     else
```

And this one ensures that the back reference to the response's request is set properly when `listen() do r::HTTP.Request ...` is used:

```patch
--- a/src/Servers.jl
+++ b/src/Servers.jl
@@ -523,6 +523,7 @@ function handle_request(f::Function, http::Stream)
     request::HTTP.Request = http.message
     request.body = read(http)
     request.response::HTTP.Response = f(request)
+    request.response.request = request
     startwrite(http)
     write(http, request.response.body)
     return
```

**Comment by samoconnor** [comment]
@quinnj, I think we should definitely make the [two changes to fix the server](https://github.com/JuliaWeb/HTTP.jl/issues/261#issuecomment-400529685). It seems fairly clear that the RFC forbids 204 Responses with `Transfer-Encoding: chunked` because a response can't both have a chunked body and "not include a message body".

As for the change to [accept these responses on the client side](https://github.com/JuliaWeb/HTTP.jl/issues/261#issuecomment-400526931), I don't see much downside. The alternative would be to stop with an error that says "The server illegally included a message body with a 204 Response". Note that section 3.3 makes a distinction between messages that do not include a body and messages with a zero length body. `0\r\n\r\n` is a zero length chunked body. A message without a body should have no  `Transfer-Encoding` or `Content-Length` headers at all.

**Comment by bauglir** [comment]
Thanks for looking into this so quickly and elaborately!

The remark on `listen` is a very insightful one. I based this example on [`handle_request`](https://github.com/JuliaWeb/HTTP.jl/blob/master/src/Servers.jl#L522) where the return value from the `handler` gets written into `request.response` as well, so I figured that was the way to do it. I tend to return a new `Response` from my handlers, I guess we should take the `request.response` and modify that one instead. Will keep an eye out for that. Although I think your second `Server` patch addresses this case right?

The cURL thing was indeed something I logged incorrectly I see.

The client-side fix fixes my issue (when applied without the server fixes), and so do the server-side fixes (when applied without the client fix)! I will leave the decision on whether to include the client-side fix up to you two :) As it looks like the server was previously behaving badly, I'd personally be fine with just those fixes. I'd probably prefer an error on the client-side, but that won't help if you don't have control over the server. So I can see the client-side fix being a lifesaver in those situations as well.
